crates/rust-analyzer/src/cli/lsif.rs

```diff
@@ -0,0 +1,260 @@
+//! Lsif generator
+
+use std::collections::HashMap;
+use std::env;
+use std::time::Instant;
+
+use ide::{
+    Analysis, Cancellable, FileId, FileRange, RootDatabase, StaticIndex, StaticIndexedFile,
+    TokenId, TokenStaticData,
+};
+use ide_db::LineIndexDatabase;
+
+use ide_db::base_db::salsa::{self, ParallelDatabase};
+use lsp_types::{lsif::*, Hover, HoverContents, NumberOrString};
+use project_model::{CargoConfig, ProjectManifest, ProjectWorkspace};
+use vfs::{AbsPathBuf, Vfs};
+
+use crate::cli::{
+    flags,
+    load_cargo::{load_workspace, LoadCargoConfig},
+    Result,
+};
+use crate::line_index::{LineEndings, LineIndex, OffsetEncoding};
+use crate::to_proto;
+
+/// Need to wrap Snapshot to provide `Clone` impl for `map_with`
+struct Snap<DB>(DB);
+impl<DB: ParallelDatabase> Clone for Snap<salsa::Snapshot<DB>> {
+    fn clone(&self) -> Snap<salsa::Snapshot<DB>> {
+        Snap(self.0.snapshot())
+    }
+}
+
+struct LsifManager<'a> {
+    count: i32,
+    token_map: HashMap<TokenId, Id>,
+    range_map: HashMap<FileRange, Id>,
+    file_map: HashMap<FileId, Id>,
+    analysis: &'a Analysis,
+    db: &'a RootDatabase,
+    vfs: &'a Vfs,
+}
+
+#[derive(Clone, Copy)]
+struct Id(i32);
+
+impl From<Id> for NumberOrString {
+    fn from(Id(x): Id) -> Self {
+        NumberOrString::Number(x)
+    }
+}
+
+impl LsifManager<'_> {
+    fn new<'a>(analysis: &'a Analysis, db: &'a RootDatabase, vfs: &'a Vfs) -> LsifManager<'a> {
+        LsifManager {
+            count: 0,
+            token_map: HashMap::default(),
+            range_map: HashMap::default(),
+            file_map: HashMap::default(),
+            analysis,
+            db,
+            vfs,
+        }
+    }
+
+    fn add(&mut self, data: Element) -> Id {
+        let id = Id(self.count);
+        self.emit(&serde_json::to_string(&Entry { id: id.into(), data }).unwrap());
+        self.count += 1;
+        id
+    }
+
+    // FIXME: support file in addition to stdout here
+    fn emit(&self, data: &str) {
+        println!("{}", data);
+    }
+
+    fn get_token_id(&mut self, id: TokenId) -> Id {
+        if let Some(x) = self.token_map.get(&id) {
+            return *x;
+        }
+        let result_set_id = self.add(Element::Vertex(Vertex::ResultSet(ResultSet { key: None })));
+        self.token_map.insert(id, result_set_id);
+        result_set_id
+    }
+
+    fn get_range_id(&mut self, id: FileRange) -> Cancellable<Id> {
```

Comment:
```suggestion
    fn get_range_id(&mut self, id: FileRange) -> Id {
```

You can only get a cancellation error if you mutate the data concurrently with queries. We don't do this here, so it should be fine to unwrap all cancellations.